File: analyze_reg.py
import pandas as pd

def analyze_product_performance(file_path):
    """
    Analyzes product performance from a CSV file, identifying top and bottom products
    by year, quarter, and month.

    Args:
        file_path (str): The path to the CSV file.
    """
    df = pd.read_csv(file_path)
    df['Date'] = pd.to_datetime(df['Date'])
    df['Year'] = df['Date'].dt.year
    df['Quarter'] = df['Date'].dt.quarter
    df['Month'] = df['Date'].dt.month

    def get_top_bottom_products(grouped_df):
        """
        Calculates top and bottom products from a grouped DataFrame.

        Args:
            grouped_df (pd.DataFrame): DataFrame grouped by a time period.

        Returns:
            tuple: A tuple containing the top and bottom product series.
        """
        total_actuals = grouped_df['Actuals'].sum()
        grouped_df['Percentage'] = (grouped_df['Actuals'] / total_actuals) * 100
        top_product = grouped_df.loc[grouped_df['Actuals'].idxmax()]
        bottom_product = grouped_df.loc[grouped_df['Actuals'].idxmin()]
        return top_product, bottom_product

    # --- Year ---
    print("--- Analysis by Year ---")
    yearly_groups = df.groupby(['Year', 'Product']).agg({'Actuals': 'sum'}).reset_index()
    for year, group in yearly_groups.groupby('Year'):
        print(f"\nYear: {year}")
        top, bottom = get_top_bottom_products(group)
        print(f"  Top Product: {top['Product']} - Amount: ${top['Actuals']:,.2f}, Percentage: {top['Percentage']:.2f}%")
        print(f"  Bottom Product: {bottom['Product']} - Amount: ${bottom['Actuals']:,.2f}, Percentage: {bottom['Percentage']:.2f}%")


    # --- Quarter ---
    print("\n--- Analysis by Quarter ---")
    quarterly_groups = df.groupby(['Year', 'Quarter', 'Product']).agg({'Actuals': 'sum'}).reset_index()
    for (year, quarter), group in quarterly_groups.groupby(['Year', 'Quarter']):
        print(f"\nQuarter: Q{quarter} {year}")
        top, bottom = get_top_bottom_products(group)
        print(f"  Top Product: {top['Product']} - Amount: ${top['Actuals']:,.2f}, Percentage: {top['Percentage']:.2f}%")
        print(f"  Bottom Product: {bottom['Product']} - Amount: ${bottom['Actuals']:,.2f}, Percentage: {bottom['Percentage']:.2f}%")


    # --- Month ---
    print("\n--- Analysis by Month ---")
    monthly_groups = df.groupby(['Year', 'Month', 'Product']).agg({'Actuals': 'sum'}).reset_index()
    for (year, month), group in monthly_groups.groupby(['Year', 'Month']):
        print(f"\nMonth: {month}/{year}")
        top, bottom = get_top_bottom_products(group)
        print(f"  Top Product: {top['Product']} - Amount: ${top['Actuals']:,.2f}, Percentage: {top['Percentage']:.2f}%")
        print(f"  Bottom Product: {bottom['Product']} - Amount: ${bottom['Actuals']:,.2f}, Percentage: {bottom['Percentage']:.2f}%")


if __name__ == "__main__":
    analyze_product_performance("F:/GEMINI/Projects/beautiful/Sample Data/REG.csv")

File: tools/analyze_reg.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var path = require("path");
var Papa = require("papaparse");
var date_fns_1 = require("date-fns");
var readline = require("readline");
var rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});
function printAnalysis(periodType, data, topN, bottomN) {
    console.log("\n--- Analysis by ".concat(periodType, " (Top ").concat(topN, ", Bottom ").concat(bottomN, ") ---"));
    var _loop_1 = function (period) {
        var products = data[period];
        var totalActuals = Object.values(products).reduce(function (sum, val) { return sum + val; }, 0);
        var sortedProducts = Object.entries(products)
            .map(function (_a) {
            var productName = _a[0], actuals = _a[1];
            return ({
                Product: productName,
                Actuals: actuals,
                Percentage: (actuals / totalActuals) * 100,
            });
        })
            .sort(function (a, b) { return b.Actuals - a.Actuals; });
        console.log("\n".concat(periodType, ": ").concat(period));
        var top_1 = sortedProducts.slice(0, topN);
        var bottom = sortedProducts.slice(-bottomN).reverse();
        console.log("  Top ".concat(topN, " Products:"));
        top_1.forEach(function (p) {
            var _a;
            console.log("    - ".concat(p.Product, ": ").concat(p.Actuals.toFixed(2), " (").concat((_a = p.Percentage) === null || _a === void 0 ? void 0 : _a.toFixed(2), "%)"));
        });
        console.log("  Bottom ".concat(bottomN, " Products:"));
        bottom.forEach(function (p) {
            var _a;
            console.log("    - ".concat(p.Product, ": ").concat(p.Actuals.toFixed(2), " (").concat((_a = p.Percentage) === null || _a === void 0 ? void 0 : _a.toFixed(2), "%)"));
        });
    };
    for (var period in data) {
        _loop_1(period);
    }
}
function startInteractiveSession(analysisData, currentPeriod, currentTopN, currentBottomN) {
    printAnalysis(currentPeriod, analysisData[currentPeriod], currentTopN, currentBottomN);
    rl.question('\nWhat would you like to do next?\n(1) Change time scale\n(2) Change Top/Bottom N\n(3) Exit\n> ', function (answer) {
        if (answer === '1') {
            rl.question('Enter new time scale (Year, Quarter, Month): ', function (scale) {
                var newScale = scale.trim().charAt(0).toUpperCase() + scale.trim().slice(1).toLowerCase();
                if (newScale === 'Year' || newScale === 'Quarter' || newScale === 'Month') {
                    startInteractiveSession(analysisData, newScale, currentTopN, currentBottomN);
                }
                else {
                    console.log("Invalid time scale. Please try again.");
                    startInteractiveSession(analysisData, currentPeriod, currentTopN, currentBottomN);
                }
            });
        }
        else if (answer === '2') {
            rl.question('Enter Top N: ', function (topNStr) {
                var topN = parseInt(topNStr, 10);
                rl.question('Enter Bottom N: ', function (bottomNStr) {
                    var bottomN = parseInt(bottomNStr, 10);
                    if (!isNaN(topN) && !isNaN(bottomN) && topN > 0 && bottomN > 0) {
                        startInteractiveSession(analysisData, currentPeriod, topN, bottomN);
                    }
                    else {
                        console.log("Invalid number. Please enter positive integers.");
                        startInteractiveSession(analysisData, currentPeriod, currentTopN, currentBottomN);
                    }
                });
            });
        }
        else if (answer === '3') {
            rl.close();
        }
        else {
            console.log("Invalid option. Please try again.");
            startInteractiveSession(analysisData, currentPeriod, currentTopN, currentBottomN);
        }
    });
}
function launchAnalysisTool(filePath) {
    var csvFile = fs.readFileSync(filePath, 'utf8');
    var yearlyTotals = {};
    var quarterlyTotals = {};
    var monthlyTotals = {};
    Papa.parse(csvFile, {
        header: true,
        skipEmptyLines: true,
        complete: function (results) {
            for (var _i = 0, _a = results.data; _i < _a.length; _i++) {
                var row = _a[_i];
                if (!row.Date || !row.Product || !row.Actuals)
                    continue;
                try {
                    var date = (0, date_fns_1.parse)(row.Date, 'M/d/yyyy', new Date());
                    var actuals = parseFloat(row.Actuals);
                    if (isNaN(actuals))
                        continue;
                    var product = row.Product;
                    var year = (0, date_fns_1.getYear)(date).toString();
                    var quarter = "Q".concat((0, date_fns_1.getQuarter)(date), " ").concat(year);
                    var month = "".concat((0, date_fns_1.getMonth)(date) + 1, "/").concat(year);
                    if (!yearlyTotals[year])
                        yearlyTotals[year] = {};
                    yearlyTotals[year][product] = (yearlyTotals[year][product] || 0) + actuals;
                    if (!quarterlyTotals[quarter])
                        quarterlyTotals[quarter] = {};
                    quarterlyTotals[quarter][product] = (quarterlyTotals[quarter][product] || 0) + actuals;
                    if (!monthlyTotals[month])
                        monthlyTotals[month] = {};
                    monthlyTotals[month][product] = (monthlyTotals[month][product] || 0) + actuals;
                }
                catch (error) {
                    console.error("Skipping invalid date format: ".concat(row.Date));
                    continue;
                }
            }
            var analysisData = { Year: yearlyTotals, Quarter: quarterlyTotals, Month: monthlyTotals };
            startInteractiveSession(analysisData, 'Month', 2, 2);
        },
    });
}
var filePath = path.resolve('F:/GEMINI/Projects/beautiful/Sample Data/REG.csv');
launchAnalysisTool(filePath);
